cabal repl (and nix build) do not respect <package>.buildinfo

**Describe the bug**
`cabal build` successfully builds [the cabal project](https://github.com/alexoundos/cabal-buildinfo-example), which uses `build-type: Configure` with `<package>.buildinfo` and CPP definition in it.
But `cabal repl` fails:
```console
$ cabal repl
Build profile: -w ghc-9.2.8 -O1
In order, the following will be built (use -v for more details):
 - cabal-buildinfo-example-0.1.0.0 (ephemeral targets)
Preprocessing library for cabal-buildinfo-example-0.1.0.0..
GHCi, version 9.2.8: https://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling MyLib            ( src/MyLib.hs, interpreted )

src/MyLib.hs:6:14: error:
    Data constructor not in scope: MYDEF :: String
  |
6 | definition = MYDEF
  |              ^^^^^
Failed, no modules loaded.
```
Remarkably, `nix build` fails the same way!

**To Reproduce**
Steps to reproduce the behavior:

```
$ git clone https://github.com/alexoundos/cabal-buildinfo-example
$ cd cabal-buildinfo-example
$ nix develop
[devshell]$ cabal build
[devshell]$ cabal run
[devshell]$ cabal repl
```

**Expected behavior**
Generated `<project>.buildinfo` file must be respected in `cabal repl` (and `nix build`).

**System information**
 - Operating system: tested on Linux (NixOS 23.05)
 - `cabal`: v3.10.1.0
 - `ghc`: v9.2.8

**Additional context**

This issue was already reported: #4195, haskell/haskell-language-server#3780.

--- a/Cabal/src/Distribution/Simple.hs
+++ b/Cabal/src/Distribution/Simple.hs
@@ -743,8 +743,7 @@ simpleUserHooks =
 --
 -- * 'postConf' runs @.\/configure@, if present.
 --
--- * the pre-hooks 'preBuild', 'preClean', 'preCopy', 'preInst',
---   'preReg' and 'preUnreg' read additional build information from
+-- * the pre-hooks, except for pre-conf, read additional build information from
 --   /package/@.buildinfo@, if present.
 --
 -- Thus @configure@ can use local system information to generate
@@ -753,7 +752,8 @@ autoconfUserHooks :: UserHooks
 autoconfUserHooks =
   simpleUserHooks
     { postConf = defaultPostConf
-    , preBuild = readHookWithArgs buildVerbosity buildDistPref -- buildCabalFilePath,
+    , preBuild = readHookWithArgs buildVerbosity buildDistPref
+    , preRepl = readHookWithArgs replVerbosity replDistPref
     , preCopy = readHookWithArgs copyVerbosity copyDistPref
     , preClean = readHook cleanVerbosity cleanDistPref
     , preInst = readHook installVerbosity installDistPref
@@ -761,6 +761,8 @@ autoconfUserHooks =
     , preHaddock = readHookWithArgs haddockVerbosity haddockDistPref
     , preReg = readHook regVerbosity regDistPref
     , preUnreg = readHook regVerbosity regDistPref
+    , preTest = readHookWithArgs testVerbosity testDistPref
+    , preBench = readHookWithArgs benchmarkVerbosity benchmarkDistPref
     }
   where
     defaultPostConf

--- a/changelog.d/issue-9401
+++ b/changelog.d/issue-9401
@@ -1,0 +1,11 @@
+synopsis: Account for .buildinfo in repl, test, and bench
+packages: Cabal
+prs: #9440
+issues: #9401
+
+description: {
+
+Generated <project>.buildinfo is now respected in cabal repl, cabal test and cabal bench.
+
+}
+

